File: NFA.java
import java.util.*;

public class NFA {
    private Set<Integer> states;
    private Set<Character> alphabet;
    private Map<Integer, Map<Character, Set<Integer>>> transitions;
    private int startState;
    private Set<Integer> acceptStates;

    public NFA(Set<Integer> states, Set<Character> alphabet, Map<Integer, Map<Character, Set<Integer>>> transitions, int startState, Set<Integer> acceptStates) {
        this.states = states;
        this.alphabet = alphabet;
        this.transitions = transitions;
        this.startState = startState;
        this.acceptStates = acceptStates;
    }

    // 可视化NFA
    public void visualize() {
        System.out.println("NFA Visualization:");
        for (Map.Entry<Integer, Map<Character, Set<Integer>>> entry : transitions.entrySet()) {
            int state = entry.getKey();
            for (Map.Entry<Character, Set<Integer>> trans : entry.getValue().entrySet()) {
                char symbol = trans.getKey();
                for (int nextState : trans.getValue()) {
                    System.out.println(state + " --" + symbol + "--> " + nextState);
                }
            }
        }
    }

    // 扫描字符串
    public boolean scan(String input) {
        Set<Integer> currentStates = new HashSet<>();
        currentStates.add(startState);
        for (char symbol : input.toCharArray()) {
            Set<Integer> nextStates = new HashSet<>();
            for (int state : currentStates) {
                if (transitions.containsKey(state) && transitions.get(state).containsKey(symbol)) {
                    nextStates.addAll(transitions.get(state).get(symbol));
                }
            }
            currentStates = nextStates;
        }
        for (int state : currentStates) {
            if (acceptStates.contains(state)) {
                return true;
            }
        }
        return false;
    }
}

File: nfa.rs
use std::collections::{HashMap, HashSet};

pub struct NFA {
    states: HashSet<u32>,
    alphabet: HashSet<char>,
    transitions: HashMap<u32, HashMap<char, HashSet<u32>>>,
    start_state: u32,
    accept_states: HashSet<u32>,
}

impl NFA {
    pub fn new(
        states: HashSet<u32>,
        alphabet: HashSet<char>,
        transitions: HashMap<u32, HashMap<char, HashSet<u32>>>,
        start_state: u32,
        accept_states: HashSet<u32>,
    ) -> Self {
        NFA {
            states,
            alphabet,
            transitions,
            start_state,
            accept_states,
        }
    }

    // 可视化NFA
    pub fn visualize(&self) {
        println!("NFA Visualization:");
        for (state, trans) in &self.transitions {
            for (symbol, next_states) in trans {
                for next_state in next_states {
                    println!("{} --{}--> {}", state, symbol, next_state);
                }
            }
        }
    }

    // 扫描字符串
    pub fn scan(&self, input: &str) -> bool {
        let mut current_states: HashSet<u32> = HashSet::new();
        current_states.insert(self.start_state);
        for symbol in input.chars() {
            let mut next_states: HashSet<u32> = HashSet::new();
            for &state in &current_states {
                if let Some(trans) = self.transitions.get(&state) {
                    if let Some(next) = trans.get(&symbol) {
                        next_states.extend(next);
                    }
                }
            }
            current_states = next_states;
        }
        current_states.intersection(&self.accept_states).count() > 0
    }
}